exp/controllers/azureasomanagedcluster_controller.go

```diff
@@ -129,6 +133,21 @@ func (r *AzureASOManagedClusterReconciler) reconcileNormal(ctx context.Context,
 	defer done()
 	log.V(4).Info("reconciling normally")
 
+	if cluster == nil {
+		log.V(4).Info("Cluster Controller has not yet set OwnerRef")
+		return ctrl.Result{}, nil
+	}
+	if cluster.Spec.ControlPlaneRef == nil ||
+		cluster.Spec.ControlPlaneRef.APIVersion != infrav1exp.GroupVersion.Identifier() ||
+		cluster.Spec.ControlPlaneRef.Kind != infrav1exp.AzureASOManagedControlPlaneKind {
+		return ctrl.Result{}, reconcile.TerminalError(errInvalidControlPlaneKind)
+	}
+
+	needsPatch := controllerutil.AddFinalizer(asoManagedCluster, clusterv1.ClusterFinalizer)
+	if needsPatch {
+		return ctrl.Result{Requeue: true}, nil
```

Comment:
I don't see `Requeue: true` being set for any of the other `ctr.Result{}`s. Is this something that's enabled by default or is this case different?

Reply:
In all the other cases we expect some other change to trigger another reconciliation as needed. This change doesn't by default though. Adding the finalizer is basically an intermediate step that we need to persist before finishing a full reconciliation, so this is how we tell the controller to immediately re-reconcile so we do all the stuff that will come after this.

Reply:
I was a little confused by this as well. How are the changes to the modified `asoManagedCluster` function parameter object preserved for patching in the next requeue if the immediate requeue signal is sent in the return object?

Reply:
We `defer` a function in the top-level `Reconcile` to patch the object after each reconciliation loop which will persist this change, if that's what's not clear.

Reply:
The existing controllers do a separate patch inline and keep going within the same reconciliation loop, but this way forces all modification of the object to go through the same patch mechanism at the end, which I feel is slightly more idiomatic.

Reply:
defer explains it!